### ISSUE TITLE
False positive for promise-function-async within a .filter()

### ISSUE THREAD

**Issue by chrismbarr** [palantir/tslint#3580]
### Bug Report

- __TSLint version__: 5.8.0
- __TypeScript version__: 2.6.2
- __Running TSLint via__: `gulp-tslint`

#### TypeScript code being linted

```ts
public static async doesExchangeAliasExist(alias: string): Promise<boolean> {
    const text = await this.emailAdressesRows.filter((row: ElementFinder) => 
                                                   //^ - error reported here
        row.$$("td").get(1).getText().then((email: string) => email === alias))
        .first()
        .getText();

    return text.includes(alias);
}
```

with `tslint.json` configuration:

```json
"promise-function-async": true
```

#### Actual behavior

When adding `async` where the rule suggests, I now get a TS build error of:
```
(TS) Argument of type '(row: ElementFinder) => Promise<boolean>' is not assignable to parameter of type '(element: ElementFinder, index?: number | undefined) => boolean | Promise<boolean>'.
  Type 'Promise<boolean>' is not assignable to type 'boolean | Promise<boolean>'.
    Type 'Promise<boolean>' is not assignable to type 'promise.Promise<boolean>'.
      Property 'cancel' is missing in type 'Promise<boolean>'.
```

This is what my IDE (Visual Studio 2017) shows
![image](https://user-images.githubusercontent.com/463685/34006487-831e2764-e0cc-11e7-9ac6-a609bb29ed80.png)

#### Expected behavior
Not to suggest fixes that will prevent code compilation. Right now it looks like I'll have to add a rule disable here since it appears to be a bad suggestion

**Comment by aervin** [comment]
@chrismbarr I thought I'd look into this but was unable to recreate the error by adding your code sample to `promise-function-async`'s test file. Any advice?

**Comment by chrismbarr** [comment]
really?  The above code doesn't give you a linting violation for this rule?

**Comment by aervin** [comment]
Not when I added this code to its test file:
```ts
public static async doesExchangeAliasExist(alias: string): Promise<boolean> {
    const text = await this.emailAdressesRows.filter((row: ElementFinder) => 
        row.$$("td").get(1).getText().then((email: string) => email === alias))
        .first()
        .getText();

    return text.includes(alias);
}
```

However, I can recreate the TS errors in VSCode. But does that mean this is not a tslint issue?

**Comment by ajafff** [comment]
The real issue is the `.filter()` method not accepting a standard ES6 promise.

```
Type 'Promise<boolean>' is not assignable to type 'promise.Promise<boolean>'.
   Property 'cancel' is missing in type 'Promise<boolean>'.
```

In theory the rule could check if `Promise` is assignable to the contextual type and don't show an error if it doesn't.
I don't know if it's worth the effort though. In addition there is currently no usable API to check type assignability in TypeScript.

**Comment by chrismbarr** [comment]
well idk what to say, it shows as a failure for me when I run my gulp linting task, and via the "TSLint" extension for Visual Studio 2017.  The rule makes a bead suggestion here, so it appears that the only way to get around this is to ignore the rule on every occurance of this.

**Comment by ajafff** [comment]
I agree that this behavior is not ideal. Unfortunately we don't have the necessary API to detect and fix such cases. This is already being worked on in the TypeScript repo.

Until then you could bug the author of the `.filter()` method to use the correct Promise type (or fix it yourself if you are the author).

**Comment by JoshuaKGoldberg** [comment]
Closing this as external since it seems there's nothing actionable on TSLint's end. Please do post here if there's something that can be done!
